### ISSUE TITLE
Issue installing aubio using pip / python

### ISSUE THREAD

**Issue by ulyssesp** [aubio/aubio#154]
`pip install aubio` gives this error:

```
Failed to build aubio
Installing collected packages: numpy, aubio
  Running setup.py install for aubio ... error
    Complete output from command c:\python35-td\scripts\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\ulyssesp\\AppData\\Local\\Temp\\pip-build-2g9tsqq_\\aubio\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\ulyssesp\AppData\Local\Temp\pip-_tsf52vv-record\install-record.txt --single-version-externally-managed --compile --install-headers c:\python35-td\include\site\python3.5\aubio:
    running install
    running build
    running build_py
    running build_ext
    checking for aubio = 0.4.6
    Running "pkg-config --libs --cflags aubio = 0.4.6" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    Info: aubio 0.4.6 was not found by pkg-config
    Info: looking for *optional* additional packages
    checking for libavcodec
    Running "pkg-config --libs --cflags libavcodec" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for libavformat
    Running "pkg-config --libs --cflags libavformat" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for libavutil
    Running "pkg-config --libs --cflags libavutil" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for libswresample
    Running "pkg-config --libs --cflags libswresample" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for libavresample
    Running "pkg-config --libs --cflags libavresample" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for sndfile
    Running "pkg-config --libs --cflags sndfile" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    checking for samplerate
    Running "pkg-config --libs --cflags samplerate" failed: FileNotFoundError(2, 'The system cannot find the file specified', None, 2, None)
    Info: libaubio was not installed or built locally with waf, adding src/
    Running command: C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe -E -x c -DAUBIO_UNSTABLE=1 -Isrc src\aubio.h
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\ulyssesp\AppData\Local\Temp\pip-build-2g9tsqq_\aubio\setup.py", line 85, in <module>
        'tests': ['numpy'],
      File "c:\python35-td\lib\site-packages\setuptools\__init__.py", line 129, in setup
        return distutils.core.setup(**attrs)
      File "C:\Python35\Lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "C:\Python35\Lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "C:\Python35\Lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\python35-td\lib\site-packages\setuptools\command\install.py", line 61, in run
        return orig.install.run(self)
      File "C:\Python35\Lib\distutils\command\install.py", line 539, in run
        self.run_command('build')
      File "C:\Python35\Lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "C:\Python35\Lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "C:\Python35\Lib\distutils\command\build.py", line 135, in run
        self.run_command(cmd_name)
      File "C:\Python35\Lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "C:\Python35\Lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "C:\Python35\Lib\distutils\command\build_ext.py", line 339, in run
        self.build_extensions()
      File "C:\Python35\Lib\distutils\command\build_ext.py", line 448, in build_extensions
        self._build_extensions_serial()
      File "C:\Python35\Lib\distutils\command\build_ext.py", line 473, in _build_extensions_serial
        self.build_extension(ext)
      File "C:\Users\ulyssesp\AppData\Local\Temp\pip-build-2g9tsqq_\aubio\python\lib\moresetuptools.py", line 192, in build_extension
        usedouble=enable_double)
      File "C:\Users\ulyssesp\AppData\Local\Temp\pip-build-2g9tsqq_\aubio\python\lib\gen_external.py", line 238, in generate_external
        c_declarations = get_c_declarations(header, usedouble=usedouble)
      File "C:\Users\ulyssesp\AppData\Local\Temp\pip-build-2g9tsqq_\aubio\python\lib\gen_external.py", line 108, in get_c_declarations
        raise Exception("preprocessor output is empty:\n%s" % err_output)
    Exception: preprocessor output is empty:
    b''
```

And `python setup.py install` gives this error:

```
aubiomodule.c
c:\users\ulyssesp\development\aubio\python\ext\aubio-types.h(4): fatal error C1083: Cannot open include file: 'aubio-generated.h': No such file or directory
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\cl.exe' failed with exit status 2
```

Any ideas as to what's going on or how to fix it?

**Comment by piem** [comment]
hello @ulyssesp 

to compile a python extension on windows, you will need the compiler that was used to build your python binary.

See this page about windows compiler to find the version appropriate to use: https://wiki.python.org/moin/WindowsCompilers

let us know if this helps!

best, piem

**Comment by ulyssesp** [comment]
Thanks for the link!

I've installed the C package from source using the 0.4.6 tag and the recommended binaries from the site. I'm still getting  the two errors shown above. It looks like part of the problem is that pkg-config can't find the aubio installed from source. Is there a way that I can point it to the correct install directory to find those libraries?

**Comment by piem** [comment]
hi @ulyssesp 

`pkg-config` is not yet available on windows at this time it seems, but that shouldn't be a problem: all of these dependencies are optional.

the problem appears to be that the output of the preprocessor is empty, so i still suspect that it's trying to use a different compiler. Maybe comparing the output you get with the [appveyor build logs](https://ci.appveyor.com/project/piem/aubio/build/1.0.1.578) could help. 

let us know how it goes,
best, piem

**Comment by 907Tyler** [comment]
Hey @ulyssesp , I imagine I'm a bit late, but I just ran into this issue as well. If you have conda installed, you can try `conda install -c conda-forge aubio` as described here https://anaconda.org/conda-forge/aubio
This allowed me to install successfully.

**Comment by Rybak5611** [comment]
Same issue

**Comment by piem** [comment]
hi @Rybak5611 

as mentioned above, you *must* have [a compiler](https://wiki.python.org/moin/WindowsCompilers) installed to build aubio's python module.

if you do not want to install a compiler, you can use [conda packages](https://anaconda.org/conda-forge/aubio).

let us know how it goes,
best, piem

**Comment by Rybak5611** [comment]
@piem I do have a compiler, in fact i have 2 of them. First i tried installing aubio with default VS compiler,  wich resulted in error above. Then i switched to MinGW by setting ``compiler=mingw32`` in distutils.cfg, but compiling with MinGW results in ``ValueError: Unknown MS Compiler version 1914``

**Comment by piem** [comment]
Hi @Rybak5611 

Thanks for the feedback.

The [python wiki](https://wiki.python.org/moin/WindowsCompilers#Which_Microsoft_Visual_C.2B-.2B-_compiler_to_use_with_a_specific_Python_version_.3F) mentions *you need to install the compiler version that corresponds to your Python version*. This is also specified in the official documentation on [building extensions C and C++ on windows](https://docs.python.org/3.7/extending/windows.html): *You will still need the C compiler that was used to build Python; typically Microsoft Visual C++*.

So we need to figure out:
 - what version of python is being used?
 - what compiler corresponds to that version of python?
 - if/once installed, why is that compiler not found by distutils?

For the last step, the [windows build logs](https://ci.appveyor.com/project/piem/aubio) might give you some hints. Note installing `wheel` and upgrading `pip` and `setuptools` was required there. Something like this should do:

     python.exe -m pip install --upgrade pip setuptools wheel

If that doesn't help, you could try building a simple C extension (aubio is not the simplest because the pre-processor is used in a first step to generate some of the C files).

Let us know how it goes!

thanks, piem

**Comment by piem** [comment]
Hi @ulyssesp and all,

This should be finally fixed with 0a8f63f.

The problem was that the (cross) compiler used on amd64, `x86_amd64/cl.exe`, produces no output when used with `-E` as a pre-processor. Using the native compiler seems to be enough.

See also #199, #208, and 062eaf1.

Thanks, piem
